torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp

```diff
@@ -17,171 +18,881 @@ namespace fuser {
+            partial_projected_extent_[merge->inner()] =
+                getMaybePartialMappedExtent(merge->out());
+          }
+        }
+        // If any domain is partially mapped, we need to clear to the left of it
+        ids.erase(ids.begin(), ids.begin() + out_pos);
+      }
     } else {
-      auto id_inner = merged_id;
-      auto id_outer = id;
-      merged_id = IterDomain::merge(id_outer, id_inner);
-      is_merge_done = true;
-    }
-  }
-  return is_merge_done ? merged_id : nullptr;
+      TORCH_INTERNAL_ASSERT(
+          false,
+          "ProjectDimensions does not support expr type: ",
+          expr->toString());
+    } // switch on expr type
+  } // For loop on the transform expressions
+
+  // Add to our tracking and return the iterator to the inserted entry
+  return projected_root_ids_.emplace(ref, ids).first;
 }
 
-namespace {
-//! Collect maximum vectorization word size of a tensor whose
-//! innermost domain is leaf_merged_domain. Contig merging is taken
-//! into account to expand vectorization if possible.
-size_t collectMaxVectorizeSizeWithContigMerge(
-    TensorView* tv,
-    IterDomain* leaf_merged_domain,
-    size_t max_vector_size_in_byte,
-    ExpressionEvaluator& expression_evaluator,
-    DataType index_type) {
-  auto dtype_size = dataTypeSize(tv->dtype(), index_type);
-  const size_t max_vector_size = max_vector_size_in_byte / dtype_size;
-
-  // Assume no halo-related expression appears in the fusion. No
-  // broadcast is merged, so indexability can be assumed to be true.
-  // This is expensive, as ContigIDs builds other things like CAMap,
-  // HaloInfo, and ConcreteBroadcast info. We should explicitly build and reuse
-  // these as they're compile time information.
-  ContigIDs contigIds(
-      {leaf_merged_domain},
-      tv->getMaybeRFactorDomain(),
-      tv->domain()->contiguity(),
-      {},
-      {},
-      getAllDivisibleSplits(tv->fusion()),
-      {},
-      true);
-
-  auto innermost_root_id = tv->getMaybeRFactorDomain().back();
-  auto indexed_id = contigIds.rootToIndexedID().at(innermost_root_id);
-
-  size_t merged_size = 1;
-  // If the indexed ID is a contig merged domain, i.e., it is
-  // different from innermost_root_id, we accumulate the extents of
-  // all the root domains covered by the contig indexed ID. Otherwise,
-  // just look at the extent of the innermost root ID.
-  if (indexed_id != innermost_root_id) {
-    const auto& within_root = contigIds.withinContigIDs().at(indexed_id);
-    for (auto root_id : tv->getMaybeRFactorDomain()) {
-      if (within_root.find(root_id) == within_root.end()) {
+// This function is very similar to projectIdToRoot, we just generally swap the
+// logic of split and merge as the reverse mapping of merge looks a lot like
+// split and vice versa.
+std::unordered_map<TensorView*, std::vector<IterDomain*>>::iterator
+ContiguousInnerDimensionsMapper::projectIdToRFactor(
+    TensorView* ref,
+    std::vector<IterDomain*> ids) {
```

Comment:
I will not read this for now. I think all my reviews for `projectIdToRoot` applies to this symmetrically. Will come back and read this after these reviews are resolved.

Reply:
Me too